// angular2localization/esm/src/services/locale.js
/**
 * ANGULAR 2 LOCALIZATION
 * An Angular 2 library to translate messages, dates and numbers.
 * Written by Roberto Simonetti.
 * MIT license.
 * https://github.com/robisim74/angular2localization
 */
/**
 * Locale superclass.
 * Provides the methods for localization.
 *
 * Extend this class in components to provide the necessary methods for localization:
 *
 * export class AppComponent extends Locale {
 *
 *     constructor(public locale: LocaleService, public localization: LocalizationService) {
 *         super(locale, localization);
 *
 *     }
 *
 * }
 *
 * @author Roberto Simonetti
 */
export class Locale {
    constructor(locale, localization) {
        this.locale = locale;
        this.localization = localization;
    }
    // Gets the language code for the LocalizationService.
    get lang() {
        return this.localization.languageCode;
    }
    // Gets the default locale.
    get defaultLocale() {
        return this.locale.getDefaultLocale();
    }
    // Gets the current currency.
    get currency() {
        return this.locale.getCurrentCurrency();
    }
}
