// angular2localization/esm/src/services/locale-parser.js
/**
 * ANGULAR 2 LOCALIZATION
 * An Angular 2 library to translate messages, dates and numbers.
 * Written by Roberto Simonetti.
 * MIT license.
 * https://github.com/robisim74/angular2localization
 */
import { isPresent, isBlank, NumberWrapper, RegExpWrapper } from '@angular/common/src/facade/lang';
import { BaseException } from '@angular/common/src/facade/exceptions';
import { NumberFormatStyle, NumberFormatter } from '@angular/common/src/facade/intl';
import { IntlSupport } from './Intl-support';
/**
 * LocaleParser class.
 * Parses a string and returns a number by default locale.
 *
 * @author Roberto Simonetti
 */
export class LocaleParser {
    /**
     * Builds the regular expression for a number by default locale.
     *
     * @param defaultLocale The default locale
     * @param digits The digit info: {minIntegerDigits}.{minFractionDigits}-{maxFractionDigits}
     * @return A RegExp object
     */
    static NumberRegExpFactory(defaultLocale, digits) {
        // Gets digits.
        var minInt = 1;
        var minFraction = 0;
        var maxFraction = 3;
        const NUMBER_FORMAT_REGEXP = /^(\d+)?\.((\d+)(\-(\d+))?)?$/g;
        if (isPresent(digits)) {
            var parts = RegExpWrapper.firstMatch(NUMBER_FORMAT_REGEXP, digits);
            if (isBlank(parts)) {
                throw new BaseException(`${digits} is not a valid digit info for number`);
            }
            if (isPresent(parts[1])) {
                minInt = NumberWrapper.parseIntAutoRadix(parts[1]);
            }
            if (isPresent(parts[3])) {
                minFraction = NumberWrapper.parseIntAutoRadix(parts[3]);
            }
            if (isPresent(parts[5])) {
                maxFraction = NumberWrapper.parseIntAutoRadix(parts[5]);
            }
        }
        // Converts numbers & signs to Unicode by default locale.
        var codes = new DecimalCode(defaultLocale);
        var minusSign = codes.minusSign;
        var zero = codes.numbers[0];
        var decimalSeparator = codes.decimalSeparator;
        var nine = codes.numbers[9];
        // Pattern for 1.2-2 digits: /^-?[0-9]{1,}\.[0-9]{2,2}$/
        // Unicode pattern = "^\u002d?[\u0030-\u0039]{1,}\\u002e[\u0030-\u0039]{2,2}$";
        var pattern;
        if (minFraction > 0 && maxFraction > 0) {
            pattern = "^"
                + minusSign
                + "?[" + zero + "-" + nine
                + "]{" + minInt + ",}\\"
                + decimalSeparator
                + "[" + zero + "-" + nine
                + "]{" + minFraction + "," + maxFraction
                + "}$";
        }
        else if (minFraction == 0 && maxFraction > 0) {
            // Decimal separator is optional.
            pattern = "^"
                + minusSign
                + "?[" + zero + "-" + nine
                + "]{" + minInt + ",}\\"
                + decimalSeparator
                + "?[" + zero + "-" + nine
                + "]{" + minFraction + "," + maxFraction
                + "}$";
        }
        else {
            // Integer number.
            pattern = "^"
                + minusSign
                + "?[" + zero + "-" + nine
                + "]{" + minInt + ",}$";
        }
        pattern = codes.UnicodeToChar(pattern);
        var regExp = new RegExp(pattern);
        return regExp;
        // Wonderful, it works!
    }
    /**
     * Parses a string and returns a number by default locale.
     *
     * @param s The string to be parsed
     * @param defaultLocale The default locale
     * @return A number. If the string cannot be converted to a number, returns NaN
     */
    static Number(s, defaultLocale) {
        if (s == "" || defaultLocale == "" || defaultLocale == null) {
            return null;
        }
        var codes = new DecimalCode(defaultLocale);
        return codes.parse(s);
    }
}
/**
 * NumberCode abstract superclass.
 *
 * Converts numbers to Unicode by locales.
 *
 * @author Roberto Simonetti
 */
class NumberCode {
    constructor(defaultLocale) {
        this.defaultLocale = defaultLocale;
        /**
         * Unicode for numbers from 0 to 9.
         */
        this.numbers = [];
        for (var i = 0; i <= 9; i++) {
            this.numbers.push(this.Unicode(i.toString()));
        }
        // Checks for support for Intl.
        if (IntlSupport.NumberFormat(defaultLocale) == true) {
            // Updates Unicode for numbers by default locale.
            for (var i = 0; i <= 9; i++) {
                this.numbers[i] = this.Unicode(NumberFormatter.format(i, defaultLocale, NumberFormatStyle.Decimal, {
                    minimumIntegerDigits: 1,
                    minimumFractionDigits: 0,
                    maximumFractionDigits: 0,
                    currency: null,
                    currencyAsSymbol: false
                }));
            }
        }
    }
    Unicode(c) {
        return "\\u" + this.HexEncode(c.charCodeAt(0));
    }
    HexEncode(value) {
        var hex = value.toString(16).toUpperCase();
        // With padding.
        hex = "0000".substr(0, 4 - hex.length) + hex;
        return hex;
    }
}
/**
 * DecimalCode class.
 *
 * Converts numbers & signs to Unicode by locales.
 *
 * @author Roberto Simonetti
 */
class DecimalCode extends NumberCode {
    constructor(defaultLocale) {
        super(defaultLocale);
        this.defaultLocale = defaultLocale;
        this.minusSign = this.Unicode("-");
        this.decimalSeparator = this.Unicode(".");
        // Checks for support for Intl.
        if (IntlSupport.NumberFormat(defaultLocale) == true) {
            // Updates Unicode for signs by default locale.
            var value = -0.9; // Reference value.
            var localeValue = NumberFormatter.format(value, defaultLocale, NumberFormatStyle.Decimal, {
                minimumIntegerDigits: 1,
                minimumFractionDigits: 1,
                maximumFractionDigits: 1,
                currency: null,
                currencyAsSymbol: false
            });
            // Checks Unicode character 'RIGHT-TO-LEFT MARK' (U+200F).
            var index;
            if (this.Unicode(localeValue.charAt(0)) != "\\u200F") {
                // Left to right.
                index = 0;
            }
            else {
                // Right to left.
                index = 1;
            }
            this.minusSign = this.Unicode(localeValue.charAt(index));
            this.decimalSeparator = this.Unicode(localeValue.charAt(index + 2));
        }
    }
    parse(s) {
        // Splits the String object into an array of characters.
        var characters = s.split("");
        // Builds the value.
        var value = "";
        for (let char of characters) {
            var charCode = this.Unicode(char);
            // Tries to look for the char code in numbers and signs.
            var index = this.numbers.indexOf(charCode);
            if (index != -1) {
                value += index;
            }
            else if (charCode == this.minusSign) {
                value += "-";
            }
            else if (charCode == this.decimalSeparator) {
                value += ".";
            }
            else {
                return NaN;
            }
        }
        return parseFloat(value);
    }
    UnicodeToChar(pattern) {
        return pattern.replace(/\\u[\dA-F]{4}/gi, (match) => {
            return String.fromCharCode(parseInt(match.replace(/\\u/g, ""), 16));
        });
    }
}
